### PR TITLE
Object tagging (#4754) 

### LINKED ISSUE #4754: Support User-Defined Object Metadata
_This is a draft proposal, and likely needs more polish_

**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Many stores provide the ability to associate arbitrary user-defined attributes with objects, it would be useful to expose this.

**Describe the solution you'd like**

I would like to propose a new `put_opts` call, in a similar vein to the existing `get_opts`. This would take a `PutOptions`

```
pub struct PutOptions {
    pub metadata: HashMap<String, String>
}
```

Stores that can't store metadata should return an error if passed metadata, and `ObjectMeta` should be updated to include such metadata.

Unix systems can likely make use of [xattr](https://crates.io/crates/xattr) to store user metadata

We will likely need to restrict the key names in some manner

**Describe alternatives you've considered**

**Additional context**

https://github.com/apache/arrow-rs/issues/4498 also calls for some sort of put_opts style API

https://github.com/apache/arrow-rs/issues/4753 would benefit from this functionality

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #4754.

# Rationale for this change
 

Some stores support object tagging and this is important for existing workloads. Unfortunately support for this is not supported at all in some cases, e.g. GCS, and inconsistently supported in others, e.g. Azure and AWS-compatible stores.

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: object_store/src/lib.rs [hunk 1]
```diff
@@ -1882,6 +1903,63 @@ mod tests {
         assert_eq!(meta.size, chunk_size * 2);
     }
 
+    #[cfg(any(feature = "aws", feature = "azure"))]
+    pub(crate) async fn tagging<F, Fut>(storage: &dyn ObjectStore, get_tags: F)
+    where
+        F: Fn(Path) -> Fut + Send + Sync,
+        Fut: Future<Output = Result<reqwest::Response>> + Send,
```
Comment (tustvold) [inline object_store/src/lib.rs:1910]: Honestly working out the correct combination of lifetimes and send and sync took longer than anything else in this PR... Sometimes I really dislike async...

File: object_store/src/aws/builder.rs [hunk 2]
```diff
@@ -299,6 +301,15 @@ pub enum AmazonS3ConfigKey {
...
     SkipSignature,
 
+    /// Skips tagging objects
+    ///
+    /// This can be desirable if not supported by the backing store
+    ///
+    /// Supported keys:
+    /// - `aws_skip_tagging`
+    /// - `skip_tagging`
+    SkipTagging,
```
Comment (tustvold) [inline object_store/src/aws/builder.rs:311]: I don't feel strongly about this name

Comment (thinkharderdev) [reply object_store/src/aws/builder.rs:311]: Maybe `DisableTagging`

Comment (thinkharderdev) [reply object_store/src/aws/builder.rs:311]: but I don't feel strongly either :)

**Review by thinkharderdev — APPROVED** [review]
LGTM. I'm not very familiar with Azure so can't say whether that is done correctly or not but the S3 implementation looks good.

**Pull request merged**